cpp/arcticdb/entity/protobuf_mappings.cpp

```diff
@@ -0,0 +1,140 @@
+    IndexDescriptorImpl output;
+    output.set_type(IndexDescriptor::Type(index_descriptor.kind()));
+    output.set_field_count(index_descriptor.field_count());
+    return output;
+}
+
+arcticdb::proto::descriptors::AtomKey encode_key(const AtomKey &key) {
+    arcticdb::proto::descriptors::AtomKey output;
+    util::variant_match(key.id(),
+        [&](const StringId &id) { output.set_string_id(id); },
+        [&](const NumericId &id) { output.set_numeric_id(id); });
+
+    output.set_version_id(key.version_id());
+    output.set_creation_ts(key.creation_ts());
+    output.set_content_hash(key.content_hash());
+
+    util::variant_match(key.start_index(),
+        [&](const StringId &id) { output.set_string_start(id); },
+        [&](const NumericId &id) { output.set_numeric_start(id); });
+
+    util::variant_match(key.end_index(),
+        [&](const StringId &id) { output.set_string_end(id); },
+        [&](const NumericId &id) { output.set_numeric_end(id); });
+
+    output.set_key_type(arcticdb::proto::descriptors::KeyType (int(key.type())));
+    return output;
+}
+
+AtomKey decode_key(const arcticdb::proto::descriptors::AtomKey& input) {
+    StreamId stream_id = input.id_case() == input.kNumericId ? StreamId(input.numeric_id()) : StreamId(input.string_id());
+    IndexValue index_start = input.index_start_case() == input.kNumericStart ? IndexValue(input.numeric_start()) : IndexValue(input.string_start());
+    IndexValue index_end = input.index_end_case() == input.kNumericEnd ? IndexValue(input.numeric_end() ): IndexValue(input.string_end());
+
+    return atom_key_builder()
+        .version_id(input.version_id())
+        .creation_ts(timestamp(input.creation_ts()))
+        .content_hash(input.content_hash())
+        .start_index(index_start)
+        .end_index(index_end)
+        .build(stream_id, KeyType(input.key_type()));
+}
+
+void copy_stream_descriptor_to_proto(const StreamDescriptor& desc, arcticdb::proto::descriptors::StreamDescriptor& proto) {
+    proto.set_in_bytes(desc.uncompressed_bytes());
+    proto.set_out_bytes(desc.compressed_bytes());
+    proto.set_sorted(arcticdb::proto::descriptors::SortedValue(desc.sorted()));
+    *proto.mutable_index() = index_descriptor_to_proto(desc.index()); //TODO pass as parameter
+    util::variant_match(desc.id(),
+        [&proto] (const StringId& str) { proto.set_str_id(str); },
+        [&proto] (const NumericId& n) { proto.set_num_id(n); });
+
+    proto.mutable_fields()->Clear();
+    for(const auto& field : desc.fields()) {
+        auto new_field = proto.mutable_fields()->Add();
+        new_field->set_name(std::string(field.name()));
+        new_field->mutable_type_desc()->set_dimension(static_cast<uint32_t>(field.type().dimension()));
+        set_data_type(field.type().data_type(), *new_field->mutable_type_desc());
+    }
+}
+
+arcticdb::proto::descriptors::TimeSeriesDescriptor copy_time_series_descriptor_to_proto(const TimeseriesDescriptor& tsd) {
```

Comment:
There is a logic to it, in that some TimeseriesDescriptors are mostly being created anew, whereas StreamDescriptors are being populated on already-existing objects, and I want to avoid the cost of an additional assignment